### ISSUE TITLE
Zend Mail fails on emails with "From XXX" header

### ISSUE THREAD

**Issue by balsdorf** [eventum/eventum#155]
If an email starts with a header such as "From test@example.com  Wed Mar 23 01:42:50 2016" then this exception will be thrown:

`Zend\Mail\Header\Exception\InvalidArgumentException: Invalid header name detected in /media/data/work/eventum-skysql/vendor/zendframework/zend-mail/src/Header/GenericHeader.php on line 57
`

I've attached an example email.

I can't find any reference to this in Zend Mail. 

My two solution ideas are to either filter this line out in the routing class or in MailMessage::createFromHeaderBody(). However I'm not sure if it would also impact us in other areas.
[from_nocolon.txt](https://github.com/eventum/eventum/files/186826/from_nocolon.txt)

**Comment by glensc** [comment]
From where such email comes exactly? downloaded over IMAP? i.e in what codepath in eventum it comes in

as the line you're referring to is in [MBOX](https://en.wikipedia.org/wiki/Mbox) file format definition of "here begins new email"

**Comment by balsdorf** [comment]
This came in through process_all_email.php from postfix.

On Wed, Mar 23, 2016 at 4:42 PM, Elan Ruusamäe <notifications@github.com>
wrote:

> From where such email comes exactly? downloaded over IMAP? i.e in what
> codepath in eventum it comes in
>
> as the line you're referring to is in MBOX
> <https://en.wikipedia.org/wiki/Mbox> file format definition of "here
> begins new email"
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/eventum/eventum/issues/155#issuecomment-200554696>
>



-- 
Bryan Alsdorf
MariaDB Corporation | http://mariadb.com

**Comment by glensc** [comment]
sorry, but no trace possible? currently ZF is used only when composing mails

as i see this is very special case when feeding raw mailbox, the `From ` stripping belongs to routing, everywhere else should be RFC822 compatible mails. but then again, need to see some trace how this kicked in, as if i read the code then i do not see how `$full_email` reaches to ZF class...

**Comment by glensc** [comment]
heh, [Mime_Helper::decode](https://github.com/eventum/eventum/blob/v3.0.10/lib/eventum/class.routing.php#L34) decodes this email as header named `from test@example.com  wed mar 23 01` with value `42:50 2016`:

```php
stdClass::__set_state(array(
   'headers' => array (
    'from test@example.com  wed mar 23 01' => '42:50 2016',
    'mime-version' => '1.0',
    'received' => 'by 10.66.139.163 with HTTP; Tue, 22 Mar 2016 18:42:41 -0700 (PDT)',
    'in-reply-to' => '<eventum.md5.5655vch1q.1tgcre8w06zsb@support.example.com>',
    'references' => '<eventum.md5.5655ru8zx.whgzm03a9m0f@support.example.com> <eventum.md5.5655vch1q.1tgcre8w06zsb@support.example.com>',
    'date' => 'Tue, 22 Mar 2016 20:42:41 -0500',
    'message-id' => '<CAAaem7dhWh5psQqmsEQ5r=KxYUxCHEgamD9yt9dtmHhfKNHgew@mail.gmail.com>',
    'subject' => 'Re: [#9936] Re: Testing 2016-03-22',
    'from' => 'Test User <test@example.com>',
    'to' => 'Eventum <issue-9936@support.example.com>',
    'content-type' => 'text/plain; charset=UTF-8',
  ),
   'ctype_primary' => 'text',
   'ctype_secondary' => 'plain',
   'ctype_parameters' =>  array (
    'charset' => 'UTF-8',
  ),
))
```

**Comment by glensc** [comment]
even i proposed PR as the fix (merge if it's ok to you). i'd like to see trace to understand the code path you get the exception

**Comment by glensc** [comment]
btw, what di you mean by this sentence?

> I can't find any reference to this in Zend Mail.

**Comment by balsdorf** [comment]
This is the trace. You need to have the incoming email trigger an outgoing email.

app.ERROR: Uncaught Exception Zend\Mail\Header\Exception\InvalidArgumentException: "Invalid header name detected" at /media/data/work/eventum/master/vendor/zendframework/zend-mail/src/Header/GenericHeader.php line 57 {"exception":"[object] (Zend\\Mail\\Header\\Exception\\InvalidArgumentException(code: 0): Invalid header name detected at /media/data/work/eventum/master/vendor/zendframework/zend-mail/src/Header/GenericHeader.php:57)"} {"file":null,"line":null,"class":null,"function":null,"memory_usage":"7 MB","memory_peak_usage":"7 MB","version":"3.0.11-dev","usr_id":"1"}                                                                                      
sh: 1: /usr/sbin/sendmail: not found                                                                                       
PHP Fatal error:  Uncaught exception 'Zend\Mail\Header\Exception\InvalidArgumentException' with message 'Invalid header name detected' in /media/data/work/eventum/master/vendor/zendframework/zend-mail/src/Header/GenericHeader.php:57
Stack trace:                                                                                                                      
#0 /media/data/work/eventum/master/vendor/zendframework/zend-mail/src/Header/GenericHeader.php(35): Zend\Mail\Header\GenericHeader::splitHeaderLine('From test@examp...')
#1 /media/data/work/eventum/master/vendor/zendframework/zend-mail/src/Headers.php(236): Zend\Mail\Header\GenericHeader::fromString('From test@examp...')
#2 /media/data/work/eventum/master/vendor/zendframework/zend-mail/src/Headers.php(200): Zend\Mail\Headers->addHeaderLine('From test@examp...', '42:50 2016')
#3 /media/data/work/eventum/master/vendor/zendframework/zend-mail/src/Storage/Part.php(103): Zend\Mail\Headers->addHeaders(Array)                    
#4 /media/data/work/eventum/master/vendor/zendframework/zend-mail/src/Storage/Message.php(54): Zend\Mail\Storage\Part->__construct(Array)                  
#5 /media/data/work/eventum/master/src/Mail/Mail in /media/data/work/eventum/master/vendor/zendframework/zend-mail/src/Header/GenericHeader.php on line 57          
PHP Stack trace:                                                                                                                                                              
PHP   1. Monolog\ErrorHandler->handleException() /media/data/work/eventum/master/vendor/monolog/monolog/src/Monolog/ErrorHandler.php:0

**Comment by balsdorf** [comment]
I looked through the Zend Mail code and couldn't find anything attempting
to deal with it. I also couldn't find any open bugs related to it.

btw, what di you mean by this sentence?
>
> I can't find any reference to this in Zend Mail.
>
>

**Comment by glensc** [comment]
yeah, the trace doesn't give answer from where in Eventum code it tries to create MailMessage object...

**Comment by balsdorf** [comment]
This is the full backtrace. 

[155-backtrace.txt](https://github.com/eventum/eventum/files/188907/155-backtrace.txt)

Here is the abbreviated one:
#0  Eventum\Mail\MailMessage::createFromHeaderBody() called at [/media/data/work/eventum/master/lib/eventum/class.mail_queue.php:43]
#1  Mail_Queue::addMail() called at [/media/data/work/eventum/master/lib/eventum/class.mail_helper.php:589]
#2  Mail_Helper->send() called at [/media/data/work/eventum/master/lib/eventum/class.notification.php:1033]
#3  Notification::notifySubscribers() called at [/media/data/work/eventum/master/lib/eventum/class.notification.php:889]
#4  Notification::notify() called at [/media/data/work/eventum/master/lib/eventum/class.note.php:409]
#5  Note::insertNote() called at [/media/data/work/eventum/master/lib/eventum/class.support.php:2596]
#6  Support::blockEmailIfNeeded() called at [/media/data/work/eventum/master/lib/eventum/class.routing.php:249]
#7  Routing::route_emails() called at [/media/data/work/eventum/master/lib/eventum/class.routing.php:75]
#8  Routing::route() called at [/media/data/work/eventum/master/src/Command/ProcessMailCommand.php:37]
#9  Eventum\Command\ProcessMailCommand->execute() called at [/media/data/work/eventum/master/src/Command/Command.php:43]
#10 Eventum\Command\Command->run() called at [/media/data/work/eventum/master/bin/process_all_emails.php:17]
#0  Eventum\Mail\MailMessage::createFromHeaderBody() called at [/media/data/work/eventum/master/lib/eventum/class.mail_queue.php:43]
#1  Mail_Queue::addMail() called at [/media/data/work/eventum/master/lib/eventum/class.mail_helper.php:589]
#2  Mail_Helper->send() called at [/media/data/work/eventum/master/lib/eventum/class.notification.php:1033]
#3  Notification::notifySubscribers() called at [/media/data/work/eventum/master/lib/eventum/class.notification.php:889]
#4  Notification::notify() called at [/media/data/work/eventum/master/lib/eventum/class.note.php:409]
#5  Note::insertNote() called at [/media/data/work/eventum/master/lib/eventum/class.support.php:2596]
#6  Support::blockEmailIfNeeded() called at [/media/data/work/eventum/master/lib/eventum/class.routing.php:249]
#7  Routing::route_emails() called at [/media/data/work/eventum/master/lib/eventum/class.routing.php:75]
#8  Routing::route() called at [/media/data/work/eventum/master/src/Command/ProcessMailCommand.php:37]
#9  Eventum\Command\ProcessMailCommand->execute() called at [/media/data/work/eventum/master/src/Command/Command.php:43]
#10 Eventum\Command\Command->run() called at [/media/data/work/eventum/master/bin/process_all_emails.php:17]
#0  Eventum\Mail\MailMessage::createFromHeaderBody() called at [/media/data/work/eventum/master/lib/eventum/class.mail_queue.php:43]
#1  Mail_Queue::addMail() called at [/media/data/work/eventum/master/lib/eventum/class.mail_helper.php:589]
#2  Mail_Helper->send() called at [/media/data/work/eventum/master/lib/eventum/class.notification.php:1033]
#3  Notification::notifySubscribers() called at [/media/data/work/eventum/master/lib/eventum/class.notification.php:889]
#4  Notification::notify() called at [/media/data/work/eventum/master/lib/eventum/class.note.php:409]
#5  Note::insertNote() called at [/media/data/work/eventum/master/lib/eventum/class.support.php:2596]
#6  Support::blockEmailIfNeeded() called at [/media/data/work/eventum/master/lib/eventum/class.routing.php:249]
#7  Routing::route_emails() called at [/media/data/work/eventum/master/lib/eventum/class.routing.php:75]
#8  Routing::route() called at [/media/data/work/eventum/master/src/Command/ProcessMailCommand.php:37]
#9  Eventum\Command\ProcessMailCommand->execute() called at [/media/data/work/eventum/master/src/Command/Command.php:43]
#10 Eventum\Command\Command->run() called at [/media/data/work/eventum/master/bin/process_all_emails.php:17]

**Comment by glensc** [comment]
so, it happened when email is not allowed to be mailed and it tries to send the original email as blocked note.

**Comment by glensc** [comment]
the fix to this issue came via #156
